File: src/LuaCP.Test/Passes/ReferenceToValueTest.cs
using System;
using NUnit.Framework;
using LuaCP.IR.Components;
using LuaCP.IR.Instructions;
using System.Linq;
using LuaCP.Debug;

namespace LuaCP.Passes
{
	[TestFixture]
	public class ReferenceToValueTest
	{
		/// <summary>
		/// Test an expression in the form:
		/// <code>
		/// x = nil
		/// if(x) x = 1
		/// else x = 2
		/// end
		/// </code>
		/// </summary>
		[Test]
		public void TestTernaryStyle()
		{
			Module module = new Module();
			ConstantPool pool = module.Constants;
			Function func = module.EntryPoint;

			Block start = func.EntryPoint;
			Block setA = new Block(func), setB = new Block(func), end = new Block(func);

			// local x; if(1) { } else { }
			ReferenceNew x = start.AddLast(new ReferenceNew(pool.Nil));
			start.AddLast(new BranchCondition(pool[1], setA, setB));

			// x = 2
			setA.AddLast(new ReferenceSet(x, pool[2]));
			setA.AddLast(new Branch(end));

			// x = 2
			setB.AddLast(new ReferenceSet(x, pool[3]));
			setB.AddLast(new Branch(end));

			// return x
			ReferenceGet getter = end.AddLast(new ReferenceGet(x));
			TupleNew tuple = end.AddLast(new TupleNew(new [] { getter }, pool.Nil));
			end.AddLast(new Return(tuple));

			new Exporter(Console.Out).ModuleLong(module);

			ReferenceToValue.Runner(func);

			Assert.AreEqual(1, end.PhiNodes.Count);

			Phi phi = end.PhiNodes.First();

			Assert.Contains(setA, phi.Source.Keys.ToList());
			Assert.Contains(setB, phi.Source.Keys.ToList());

			Assert.AreEqual(pool[2], phi.Source[setA]);
			Assert.AreEqual(pool[3], phi.Source[setB]);
		}

		/// <summary>
		/// Test an expression in the form:
		/// <code>
		/// x = 2
		/// if (1) x = 3
		/// else ()
		/// return x
		/// </code>
		/// </summary>
		[Test]
		public void TestLongOrStyle()
		{
			Module module = new Module();
			ConstantPool pool = module.Constants;
			Function func = module.EntryPoint;

			Block start = func.EntryPoint;
			Block set = new Block(func), end = new Block(func), empty = new Block(func);

			// local x; if(1) { } else { }
			ReferenceNew x = start.AddLast(new ReferenceNew(pool[2]));
			start.AddLast(new BranchCondition(pool[1], empty, set));

			// x = 3
			set.AddLast(new ReferenceSet(x, pool[3]));
			set.AddLast(new Branch(end));

			// Nothing here
			empty.AddLast(new Branch(end));

			// return x
			ReferenceGet getter = end.AddLast(new ReferenceGet(x));
			TupleNew tuple = end.AddLast(new TupleNew(new [] { getter }, pool.Nil));
			end.AddLast(new Return(tuple));

			new Exporter(Console.Out).ModuleLong(module);

			ReferenceToValue.Runner(func);

			Assert.AreEqual(1, end.PhiNodes.Count);

			Phi phi = end.PhiNodes.First();

			Assert.Contains(empty, phi.Source.Keys.ToList());
			Assert.Contains(set, phi.Source.Keys.ToList());

			Assert.AreEqual(pool[2], phi.Source[empty]);
			Assert.AreEqual(pool[3], phi.Source[set]);
		}

		/// <summary>
		/// Test an expression in the form:
		/// <code>
		/// x = 2
		/// if(1) x = 3
		/// return x
		/// </code>
		/// </summary>
		[Test]
		public void TestLonghandOrStyle()
		{
			Module module = new Module();
			ConstantPool pool = module.Constants;
			Function func = module.EntryPoint;

			Block start = func.EntryPoint;
			Block set = new Block(func), end = new Block(func);

			// local x; if(1) { } else { }
			ReferenceNew x = start.AddLast(new ReferenceNew(pool[2]));
			start.AddLast(new BranchCondition(pool[1], end, set));

			// x = 3
			set.AddLast(new ReferenceSet(x, pool[3]));
			set.AddLast(new Branch(end));

			// return x
			ReferenceGet getter = end.AddLast(new ReferenceGet(x));
			TupleNew tuple = end.AddLast(new TupleNew(new [] { getter }, pool.Nil));
			end.AddLast(new Return(tuple));

			new Exporter(Console.Out).ModuleLong(module);

			ReferenceToValue.Runner(func);

			Assert.AreEqual(1, end.PhiNodes.Count);

			Phi phi = end.PhiNodes.First();

			Assert.Contains(start, phi.Source.Keys.ToList());
			Assert.Contains(set, phi.Source.Keys.ToList());

			Assert.AreEqual(pool[2], phi.Source[start]);
			Assert.AreEqual(pool[3], phi.Source[set]);
		}
	}
}

File: src/LuaCP.Test/Passes/ReferenceToValueTest.fs
module LuaCP.Passes.ReferenceToValueTest

open NUnit.Framework
open LuaCP.IR.Components
open LuaCP.IR.Instructions
open System.Linq

[<Test>]
let ``Ternary style: x = nil, if 1 then x = 2 else x = 3 end``() = 
    let modu = Module()
    let pool = modu.Constants
    let func = modu.EntryPoint
    // Blocks
    let start = func.EntryPoint
    let setA, setB, finish = Block(func), Block(func), Block(func)
    // local x; if(1) { } else { }
    let x = start.AddLast(ReferenceNew pool.Nil)
    start.AddLast(BranchCondition(pool.[1], setA, setB)) |> ignore
    // x = 2
    setA.AddLast(ReferenceSet(x, pool.[2])) |> ignore
    setA.AddLast(Branch(finish)) |> ignore
    // x = 3
    setB.AddLast(ReferenceSet(x, pool.[3])) |> ignore
    setB.AddLast(Branch(finish)) |> ignore
    // return x
    let getter = finish.AddLast(ReferenceGet x)
    let tuple = finish.AddLast(TupleNew([| getter |], pool.Nil))
    finish.AddLast(Return(tuple)) |> ignore
    // Run pass
    Assert.IsTrue(PassManager(modu).RunPass(ReferenceToValue.Runner, func), "Expected successful pass")
    // Check phi nodes
    Assert.AreEqual(1, finish.PhiNodes.Count)
    let phi = finish.PhiNodes.First()
    Assert.Contains(setA, phi.Source.Keys.ToList())
    Assert.Contains(setB, phi.Source.Keys.ToList())
    Assert.AreEqual(pool.[2], phi.Source.[setA])
    Assert.AreEqual(pool.[3], phi.Source.[setB])

[<Test>]
let ``Long or: x = 3 if 1 then x = 2 else end``() = 
    let modu = Module()
    let pool = modu.Constants
    let func = modu.EntryPoint
    // Blocks
    let start = func.EntryPoint
    let setA, empty, finish = Block(func), Block(func), Block(func)
    // local x; if(1) { } else { }
    let x = start.AddLast(ReferenceNew pool.[3])
    start.AddLast(BranchCondition(pool.[1], setA, empty)) |> ignore
    // x = 2
    setA.AddLast(ReferenceSet(x, pool.[2])) |> ignore
    setA.AddLast(Branch(finish)) |> ignore
    // Empty block
    empty.AddLast(Branch(finish)) |> ignore
    // return x
    let getter = finish.AddLast(ReferenceGet x)
    let tuple = finish.AddLast(TupleNew([| getter |], pool.Nil))
    finish.AddLast(Return(tuple)) |> ignore
    // Run pass
    Assert.IsTrue(PassManager(modu).RunPass(ReferenceToValue.Runner, func), "Expected successful pass")
    // Check phi nodes
    Assert.AreEqual(1, finish.PhiNodes.Count)
    let phi = finish.PhiNodes.First()
    Assert.Contains(setA, phi.Source.Keys.ToList())
    Assert.Contains(empty, phi.Source.Keys.ToList())
    Assert.AreEqual(pool.[2], phi.Source.[setA])
    Assert.AreEqual(pool.[3], phi.Source.[empty])

[<Test>]
let ``Short or: x = 3 if 1 then x = 2 end``() = 
    let modu = Module()
    let pool = modu.Constants
    let func = modu.EntryPoint
    // Blocks
    let start = func.EntryPoint
    let setA, finish = Block(func), Block(func)
    // local x; if(1) { } 
    let x = start.AddLast(ReferenceNew pool.[3])
    start.AddLast(BranchCondition(pool.[1], setA, finish)) |> ignore
    // x = 2
    setA.AddLast(ReferenceSet(x, pool.[2])) |> ignore
    setA.AddLast(Branch(finish)) |> ignore
    // return x
    let getter = finish.AddLast(ReferenceGet x)
    let tuple = finish.AddLast(TupleNew([| getter |], pool.Nil))
    finish.AddLast(Return(tuple)) |> ignore
    // Run pass
    Assert.IsTrue(PassManager(modu).RunPass(ReferenceToValue.Runner, func), "Expected successful pass")
    // Check phi nodes
    Assert.AreEqual(1, finish.PhiNodes.Count)
    let phi = finish.PhiNodes.First()
    Assert.Contains(setA, phi.Source.Keys.ToList())
    Assert.Contains(start, phi.Source.Keys.ToList())
    Assert.AreEqual(pool.[2], phi.Source.[setA])
    Assert.AreEqual(pool.[3], phi.Source.[start])
